### ISSUE TITLE
Allow gradient calculation in `DiffusionPipeline.__call__` 

### ISSUE THREAD

**Issue by JoaoLages** [huggingface/diffusers#529]
Nowadays the gradient calculation is disabled when calling the `DiffusionPipeline` object (example in [here](https://github.com/huggingface/diffusers/blob/c0493723f7851c4a5cc05ef55d3320bd9ab7718d/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L93)).
IMO, this makes the framework less flexible to work with some use cases. For example, [this framework](https://github.com/JoaoLages/diffusers-interpret/blob/e07e40c4faa9b3f8a129f2e71e72d3ae126439ff/src/diffusers_interpret/explainers/stable_diffusion.py#L78) had to copy the whole `__call__` method just to be able to get the gradients attached to the output image tensor.

**Comment by anton-l** [comment]
Hi @JoaoLages! I feel like `torch.no_grad` is a good default for pipelines, as they're intended only for inference by design. Additional functionality would require implementing a custom pipeline, indeed.

For your use case, however, there's a neat python trick to unwrap the nograd-decorated method and keep its contents: `pipeline.__call__ = pipeline.__call__.__wrapped__`. Could you check if it works for you?

**Comment by JoaoLages** [comment]
> Hi @JoaoLages! I feel like `torch.no_grad` is a good default for pipelines, as they're intended only for inference by design. Additional functionality would require implementing a custom pipeline, indeed.
> 
> For your use case, however, there's a neat python trick to unwrap the nograd-decorated method and keep its contents: `pipeline.__call__ = pipeline.__call__.__wrapped__`. Could you check if it works for you?

I tried that in an early stage but [it is not possible to override `__call__` afaik](https://stackoverflow.com/questions/60062100/is-it-possible-to-override-a-class-call-method)
